server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/ChangelogLoader.java

```diff
@@ -0,0 +1,197 @@
+import com.google.common.collect.Lists;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.sonar.api.rule.RuleKey;
+import org.sonar.api.server.ServerSide;
+import org.sonar.core.util.stream.Collectors;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.qualityprofile.QProfileChangeDto;
+import org.sonar.db.qualityprofile.QProfileChangeQuery;
+import org.sonar.db.rule.RuleDto;
+import org.sonar.db.user.UserDto;
+
+import static java.util.Objects.requireNonNull;
+
+@ServerSide
+public class ChangelogLoader {
+
+  private final DbClient dbClient;
+
+  public ChangelogLoader(DbClient dbClient) {
+    this.dbClient = dbClient;
+  }
+
+  /**
+   * @return non-null list of changes, by descending order of date
+   */
+  public Changelog load(QProfileChangeQuery query) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      List<QProfileChangeDto> dtos = dbClient.qProfileChangeDao().selectByQuery(dbSession, query);
+      List<Change> changes = dtos.stream()
+        .map(Change::from)
+        .collect(Collectors.toList(dtos.size()));
+      completeUserAndRuleNames(dbSession, changes);
+
+      int total = dbClient.qProfileChangeDao().countForProfileKey(dbSession, query.getProfileKey());
+      return new Changelog(total, changes);
+    }
+  }
+
+  private void completeUserAndRuleNames(DbSession dbSession, List<Change> changes) {
+    Set<String> logins = changes.stream().filter(c -> c.userLogin != null).map(c -> c.userLogin).collect(Collectors.toSet());
+    Map<String, String> userNamesByLogins = dbClient.userDao()
+      .selectByLogins(dbSession, logins)
+      .stream()
+      .collect(java.util.stream.Collectors.toMap(UserDto::getLogin, UserDto::getName));
+
+    Set<RuleKey> ruleKeys = changes.stream().filter(c -> c.ruleKey != null).map(c -> c.ruleKey).collect(Collectors.toSet());
+    Map<RuleKey, String> ruleNamesByKeys = dbClient.ruleDao()
+      .selectByKeys(dbSession, Lists.newArrayList(ruleKeys))
+      .stream()
+      .collect(java.util.stream.Collectors.toMap(RuleDto::getKey, RuleDto::getName));
+
+    changes.forEach(c -> {
+      c.userName = userNamesByLogins.get(c.userLogin);
+      c.ruleName = ruleNamesByKeys.get(c.ruleKey);
+    });
+  }
+
+  static class Change {
+    private String key;
+    private String type;
+    private long at;
+    private String severity;
+    private String userLogin;
+    private String userName;
+    private String inheritance;
+    private RuleKey ruleKey;
+    private String ruleName;
+    private final Map<String, String> params = new HashMap<>();
+
+    private Change() {
+    }
+
+    @VisibleForTesting
+    Change(String key, String type, long at, @Nullable String severity, @Nullable String userLogin, @Nullable String userName, @Nullable String inheritance, @Nullable RuleKey ruleKey, @Nullable String ruleName) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Constructor has 9 parameters, which is greater than 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00107)
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Split this 212 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00103)